### PR TITLE
Feature: Menu API

### PR DESCRIPTION
**Pull request by sgammon**
Introduces the client side of the new Menu API. Includes related work upstream in schema and platform.

Upstream work:
- bloombox/Schema#37 (*Feature: Menu API*, upstream at `feature/menu`)
- bloombox/platform#35 (*Feature: Menu API*, upstream at `feature/menu`)

Ongoing work:
- [x] Sync schema for new Menu API files
- [x] Update copyright year to 2018
- [x] API interface

### CODE REVIEW

File: src/main/kotlin/bloombox/client/services/menu/MenuClient.kt [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+ */
+@Suppress("unused", "CanBeParameter", "MemberVisibilityCanPrivate")
+class MenuClient(override val host: String,
+                 override val port: Int,
+                 override val apiKey: String,
+                 override val timeout: Duration,
+                 override val executor: Executor = Executors.newSingleThreadExecutor(),
+                 internal val defaultPartner: String? = null,
+                 internal val defaultLocation: String? = null) : RPCClient(), ServiceClient {
+  // -- Internals -- //
```
Comment (sgammon) [inline src/main/kotlin/bloombox/client/services/menu/MenuClient.kt:46]: here is the new `MenuClient` object

File: src/main/kotlin/bloombox/client/services/menu/MenuClient.kt [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+  private val future: MenuGrpc.MenuFutureStub = MenuGrpc.newFutureStub(channel)
+
+  // -- API: Retrieve Menu -- //
+  /**
+   * Fetch the entire menu for a given location.
+   */
+  @Throws(ServiceClientException::class, StatusRuntimeException::class)
+  fun retrieve(context: MenuContext = MenuContext.defaultContext(),
+               full: Boolean = false): GetMenu.Response {
+    val rendered = context.serialize(defaultPartner, defaultLocation)
```
Comment (sgammon) [inline src/main/kotlin/bloombox/client/services/menu/MenuClient.kt:146]: here is the client-side implementation of retrieving a menu

File: src/main/kotlin/bloombox/client/services/menu/MenuClientError.kt [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+
+/**
+ * Enumerates known or otherwise identifiable menu errors that might occur.
+ */
+enum class MenuClientError : ClientError {
+  /**
+   * Specifies that the partner code was missing or invalid.
+   */
+  PARTNER_INVALID {
```
Comment (sgammon) [inline src/main/kotlin/bloombox/client/services/menu/MenuClientError.kt:29]: client-side errors that might occur -> `MenuClientError`

File: src/test/kotlin/bloombox/client/menu/GetMenuTest.kt [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+  /**
+   * Test fetching basic menu.
+   */
+  @test
+  fun testGetBasicMenu() {
+    val response = client.platform.menu().retrieve(
+          MenuClient.MenuContext(
+                partner = partnerID,
+                location = locationID))
```
Comment (sgammon) [inline src/test/kotlin/bloombox/client/menu/GetMenuTest.kt:52]: example of fetching a menu, with an explicit specification of `partner`/`location` - i.e. via `MenuClient.MenuContext`

File: src/test/kotlin/bloombox/client/menu/GetMenuTest.kt [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    assertNotNull(response, "response from server for menu should not be null")
+    assertTrue(response.hasCatalog(), "response from server should specify menu data")
+  }
+
+  /**
+   * Test fetching basic menu with default context.
+   */
+  @test
+  fun testGetBasicMenuDefaultContext() {
+    val response = client.platform.menu().retrieve()
```
Comment (sgammon) [inline src/test/kotlin/bloombox/client/menu/GetMenuTest.kt:63]: example of how to use the menu client with the default `partner`/`location` - i.e. set via the main `BloomboxClient` when you construct the whole thing

**Review by sgammon — COMMENTED** [review]
lots of other small changes, including a copyright update to 2018, hence the massive PR.

**Review by tyler-porras — APPROVED** [review]

**Pull request merged**